src/main/java/impl/weather/impl/YearTemperatureStatsImpl.java

```
@@ -0,0 +1,86 @@
+        dayTemperatureInfos = new HashMap<>();
+    }
+
+    @Override
+    public void updateStats(DayTemperatureInfo newInfo) {
+        if (newInfo == null) {
+            throw new IllegalArgumentException("Null passed as update data");
+        }
+        var month = newInfo.getMonth();
+//        welp, next line sums up all java collections pretty well
+//        Why is in named COMPUTE IF ABSENT, why does it properly return existing value, why it accepts lambda?
+        var monthStats = dayTemperatureInfos.computeIfAbsent(month, m -> new ArrayList<>());
+        var existingDayInfo = monthStats.stream()
+                .filter(dayInfo -> dayInfo.getDay() == newInfo.getDay())
+                .findFirst()
+                .orElse(null);
+
+        if (existingDayInfo == null) {
+            monthStats.add(newInfo);
+        } else {
+//        picking ArrayList helps us a little here...
+            monthStats.set(monthStats.indexOf(existingDayInfo), newInfo);
+        }
+    }
+
+    @Override
+    public Double getAverageTemperature(Month month) {
+        if (!dayTemperatureInfos.containsKey(month)) {
+            return null;
+        }
+        return dayTemperatureInfos.get(month).stream()
+                .mapToDouble(DayTemperatureInfo::getTemperature)
+                .average().getAsDouble(); //Yeah, not sure how can we end up without actual value here
+    }
+
+    @Override
+    public Map<Month, Integer> getMaxTemperature() {
+        return dayTemperatureInfos.entrySet().stream()
+                .collect(
+                        Collectors.toMap(
+                                Map.Entry::getKey,
+                                v -> v.getValue().stream()
+                                        .map(DayTemperatureInfo::getTemperature)
+                                        .max(Integer::compareTo)
+                                        .get())); // Same here, not sure how can we end up with no value, so let it roll
```

use `orElseThrow` here

---

Not sure what you mean here

---

Disregard that

---

It is the invalid state when max is not presented, so you can throw an exception here using orElseThrow